Add missing @since annotations for 4.00 .. 4.05

edwintorok opened the pull request:
Fixes https://caml.inria.fr/mantis/view.php?id=7412

After finding a few missing @since annotations in the Unix module, I wrote the script below to find more (it is not guaranteed that it finds all missing annotations, but it found quite a few).

```bash
URL_PREFIX="https://github.com/ocaml/ocaml/commit"
check () {
    git diff --name-only $1..$2 -- "$3/*.mli" |\
    grep -Ev internal\|obj\|threadUnix\|raw | while read f; do
        echo "$f"
        rm -f commit_ids lines
        # find newly introduces functions/modules/exceptions/types
        git diff $1..$2 -- "$f" |\
            grep -Eo '^[+-] *(val|external|module|exception|type) [^:=*;]+'|\
            colrm 1 1|sed -re 's/[.]? *?$//'|sort|uniq -c|grep '^ *1'|\
            cut -f2 -d'1'|while read line; do
            echo " $line" >>lines
            # find the commit that introduced it
            # COMMIT=$(git log --pretty=format:%H -S "$line" -- "$f"|head -1)
            # echo "$COMMIT" >>commit_ids
        done
        if [ -f lines ]; then
            echo "$f"
            cat lines | while read line; do
                echo " $line"
                # find existing @since annotation if any
                grep "^ *$line" -A10 -- "$f" | tail -n +2| tr '\n' ' '|grep -Eo '^ *\(\*\*[^*]+@since[^*]+'|cut -f2 -d@
            done
            test -f commit_ids && sort -u commit_ids | while read id; do
                # unique commit ids for verification
                echo " $URL_PREFIX/$id";
                git tag --contains "$id" | head -1|while read tag; do
                    # @since tag based on commit (could be higher if cherry-picked and edited)
                    echo " @since $tag"
                done
            done
        fi
    done
}

VPREV=3.12.1
for V in 4.00.0 4.01.0 4.02.0 4.03.0 4.04.0 origin/trunk; do
    echo "------------------"
    echo "@since $V"
    for dir in otherlibs/* stdlib; do
        check "$VPREV" "$V" "$dir"
    done
    VPREV="$V"
done
```

alainfrisch commented:
Thanks, this is very useful.  I'm tempted to merge without checking each annotation one by one if you are reasonably confident in the methodology (and unless someone else objects to it, of course).

dra27 commented:
There's a little inconsistency of, e.g., 4.03 vs 4.03.0. Is it a good idea to put 4.05 before that's confirmed?

gasche commented on otherlibs/threads/thread.mli:
```diff
@@ -40,7 +40,8 @@ val self : unit -> t
    is an integer that identifies uniquely the thread.
-   It can be used to build data structures indexed by threads. *)
+   It can be used to build data structures indexed by threads.
+   @since 4.03 *)
```
This annotation is not quite right, `Thread.id` was added by ee63e8d356b5717a32b0739fbe957f8e92d6e2c1 back in 1996.

edwintorok replied:
indeed it used to be `external id` before, but I don't think that distinction matters for user code.

gasche commented:
This scripts looks phenomenally useful. Which raises two questions:
- Could you include it in `tools/`? We have other related scripts for quality analysis, eg. `tools/check-typo` 
- Is it possible to make it generic enough to work for user packages? The compiler distribution is far from the only library that would benefit from this. Maybe we could think of reimplementing it as an OCaml script to have a more convenient user-interface? (Not using bash is in no way mandatory, but in my personal experience as soon as I want to make something reusable I quickly need more robust abstraction facilities.)

@alainfrisch : You will call me a perfectionist, but I think that at least having a look at all functions to see if something looks odd would be warranted. I only looked at a few and found a couple errors, it would be interesting to iterate to fix the script.

gasche commented:
> There's a little inconsistency of, e.g., 4.03 vs 4.03.0.

Indeed it would be better to use correct minor/bugfix versions when possible -- although if it's too painful, I think it's fine to merge without.

> Is it a good idea to put 4.05 before that's confirmed?

In Batteries I use `@since NEXT_RELEASE` and grep and replace at release time. That said, while we don't know yet whether we will have a minor/bugfix release, I don't think that it should have any new standard library functions in it, so 4.05.0 is probably fine.

dbuenzli commented:
> Is it possible to make it generic enough to work for user packages?

Except maybe for larger projects I think that for user packages you rather want to simply use the documentation that corresponds to the smallest version of the package you code against. `odig`ging your installed packages gives you just that.

edwintorok commented:
@gasche I don't think there is any need to rush in merging this PR, I can iterate/improve the script if you want this in `tools/`, and go through the patch once more by hand. Regex based match won't be perfect, (but I don't want to turn this into an overkil by looking at `ocamlobjinfo` output or something).
@dra27 which form is preferred for @since? I've seen both used in existing documentation, I tend to prefer the full version number form.

gasche commented:
I definitely think that it would be good to have the script in `tools/`, because it would make it easier to distribute its usage (and the responsibility to use it) among contributors in the future. It's of course fine if the output is imperfect -- and for future runs the output should be much shorter, so having a human review it would not be an issue. (You could even have the script write a small message pointing out that the output is known to be imperfect, if you would feel guilty about it otherwise.) The goal of the current review/iteration should not be to have a "perfect output", but to find among the inevitable glitches all the one that you can relatively easily solve programmatically.

I don't think there are strong conventions on whether you should use `@since X.Y.Z` or just `@since X.Y`, but I think that the former should be preferred if it does not make the script noticeably more complex. In particular, I'm sure some older releases added functions in X.Y.Z that were not in available X.Y.0.

edwintorok committed:
Documentation: improve @since annotations

Add missing @since annotations for OCaml versions 4.00.0 - 4.05.0,
and fix existing annotations as needed:

Format.ikprintf: clarify ambiguity on @since 4.0 annotation
See https://github.com/ocaml/ocaml/commit/b81519668f7fa3d84f1453dc022a0d4a9b722eb8

Hashtbl.is_randomized and ListLabels.sort_uniq should be @since 4.03
List.sort_uniq is 4.02 but ListLabels.sort_uniq is 4.03
See:
https://github.com/ocaml/ocaml/commit/512d128918544ae1da0c808e811f3a7f177524d2
https://github.com/ocaml/ocaml/commit/189d29bfcf98525f63c1c3c4a4a2be9989a62ef7

edwintorok committed:
Documentation tool: add tools/lintapidiff.ml

Run 'make lintapidiff' in the root of a git checkout to get a list of
potentially missing or wrong @since annotations.

The tool is not built by default, you have to first run 'make
world.opt', and then run 'make lintapidiff'.

lintapidiff doesn't support stop comments: add explicit list of changes to ignore.

see copyright header for license.

edwintorok commented:
Sorry for the delay. I have updated my PR, changes since last time:

* rebased on latest trunk
* added `tools/lintapidiff.{ml,sh}`
  * works on a `Parsetree` (`Env.fold*` would be nice, but it is harder to extract the module annotations).  might've exagerated a bit with this, maybe this is not what you meant by 'reimplement in OCaml'
  * assumptions
    * you are running the tool on a development branch that is newer than any other tagged release (otherwise it complains that you deleted a bunch of values without marking them as deprecated first)
    * you have `OCaml` and `findlib` already installed (it is not integrated into the compiler's build system)
    * you are running the tool in a `git` checkout
    * version numbers are of form major.minor or major.minor.patch
    * if a value is marked as deprecated no further error is reported
    * @since/@deprecated annotations are considered to be "inherited" from the parent module (although when reporting you get one report for each value)
    * @since annotations can be extract from `ocaml.doc` attributes using a regex
    * missing annotations on undocumented symbols of undocumented modules are not reported
    * it analyzes all tagged releases since 3.0: takes ~11s in total, YMMV (needed to correctly report @since mismatches for now)
    * its scope is limited to that of this PR (analyzes only values, types and exceptions; reports only on missing @since for >= 4.00.0). Tried to look at type constructors too, but there are too many false positives from Bigarray's GADTs (which I assume shouldn't be all annotated)
  * only looks at newly added, deleted symbols and mismatched @since. Detecting other ways of breaking an API is out of scope.
  * running on trunk shows [these problems](https://gist.github.com/anonymous/d5d15e4ea0dce63ae21329ed8cd512ac)
* fixed the [problems reported](https://gist.github.com/anonymous/6271599572464e50e8e4b134c317cfc9) by the above tool
  * incorrect @since in my previous PR on Hashtbl.hash_param, Format.get_formatter_output_functions, Thread.id
  * add annotations for Bigarray.Array1.slice, Bigarray.genarray_of_array0, Format.formatter_out_functions, Hashtbl.*.{find_opt, filter_map_inplace,reset}, Hashtbl.{stats,statistics}, StringLabels.init, Sys.backend_type
* use major.minor.patch format for newly added @since annotations consistently

If you prefer I can put the tool into a separate PR, it can probably still be improved.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+open Location
+open Parsetree
+
+module IdMap = Depend.StringMap
```
Is there any particular reason to use `Depend.StringMap` instead of `Map.Make(String)` here? I would find the latter more clear as to what the interface is, and I don't immediately see why you chose to reference `Depend`.

edwintorok replied:
I was looking at the code of `depend.ml` to learn how to use the parsetree, I'll remove it.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+
+  let find_attr lst attrs =
+    try Some (List.find (fun (loc, _) -> List.mem loc.txt lst) attrs)
+    with Not_found -> None
```
Nitpicking, but `match List.find ... with exception Not_found -> None | attr -> Some attr` is a better style nowadays. Don't change it here if this one is more natural to you -- also having code that works on older OCaml version might make sense for this program.

edwintorok replied:
Yep, I used this once below when I needed tail recursion, but otherwise I'm still used to the old way.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+        with Not_found -> false
+
+  let get parent_info loc attrs =
+    let doc = get_doc ["ocaml.doc"; "ocaml.text"] attrs in
```
I wonder, shouldn't we also support unqualified `doc` and `text`? Are such explicitly-user-written attributes never considered by ocamldoc?

Octachron replied:
Ocamldoc does not use documentation attributes currently: their generation by the lexer is even disabled by ocamldoc since #348. (I also think it might be a good idea to support unqualified `doc` and `text`.)

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+    match get_doc ["ocaml.text"] attrs with (* for toplevel module annotation *)
+    | None -> false
+    | Some text ->
+        try Misc.search_substring "@deprecated" text 0 >= 0;
```
Nitpicking: the semicolon here is mildly confusing (usually people use it to indicate that a value is of type `unit`, not that it is ignored) and would break with `-strict-sequence`.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+      deprecated = parent_info.deprecated || is_deprecated attrs;
+      loc;
+      has_doc_parent = parent_info.has_doc;
+      has_doc = doc <> None
```
What is `has_doc_parent` used for? One thing that is a bit strange in this code is that the other attributes will behave nicely if there are several nested parents, but these two attributes will only refer to the immediate parent and not the aggregated result on all parents. In particular, if I understand correctly, a undocumented value in an undocumented module in a documented module will be consider undocumented.

edwintorok replied:
Making `has_doc_parent` fully propagated (i.e. `parent_info.has_doc_parent || parent_info.has_doc`) produces a lot of noise for MoreLabels, although I could easily just exclude morelabels from checking, and enable full propagation.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+      | Pmty_functor ({txt;_}, _, m) ->
+          let path = Path.Papply(path, Path.Pident (Ident.create txt)) in
+          add_module_type path m (inherits, map)
+      | _ -> map
```
Nitpicking: this is a fragile exhaustive pattern. I prefer to list all remaining constructors with `_` arguments (`Pmty_ident _ | Pmty_with _ | Pmty_typeof _ | Pmty_extension _ | Pmty_alias _` in this case) to be warned if a new case is introduced later.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+        | Some ty ->
+            enter_path "module type" path s.pmtd_name ty s.pmtd_attributes map
+        end
+    | _ -> map
```
Same comment on fragile exhaustive clauses.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+    try
+      let id =
+        orig |> Filename.chop_extension |> Filename.basename |>
+        String.capitalize_ascii |> Ident.create in
```
Could we have a function `Misc.modname_of_path : string -> string`? Almost all of the 19 occurrences of `String.capitalize_ascii` in the compiler codebase are doing this, in slightly different ways (you chop before basename, usually it seems the converse in done) and slightly incoherent ways (you only chop the last extension, some try to chop all extensions).

edwintorok replied:
maybe as a separate PR?

damiendoligez replied:
it would be nice as a separate PR

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+      let id =
+        orig |> Filename.chop_extension |> Filename.basename |>
+        String.capitalize_ascii |> Ident.create in
+      let ast = Pparse.file ~tool_name:Sys.argv.(0) Format.err_formatter input
```
I think `let tool_name = "lintapidiff"` would be more coherent with other tools (see `git grep "let tool_name"`). In particular `Sys.argv.(0)` is an absolute path on my machine when the tool is run from the PATH, so the difference is observable.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+  let with_show ~f rev path =
+    let obj = rev ^ ":" ^ path in
+    let suffix = Printf.sprintf "-%s:%s" rev (Filename.basename path) in
+    let tmp = Filename.temp_file "" suffix in
```
`"lintapidiff"` would be a nice prefix. These files have a tendency to end up uncleaned for various reasons (congratulations for the nice `try_finally` here) and it's nice to know where they come from.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+
+  let check_changes ~first ~last default k seen latest =
+    let is_old v =
+      version_of_string v < (4,0,0) || same_version v first
```
Relying on the field ordering of polymorphic comparison is kind of ugh -- on the other hand, writing the order by hand is a pain. You could have a `strictly_older_version` function that encapsulates that.

gasche replied:
Besides, it would be nice to get the magic constant `(4,0,0)` at the toplevel with some comment to explain what its meaning is (basically, we give up on older versions), so that people can more easily tweak that part of the behavior.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+    deprecated: bool;
+  }
+
+  let err k (loc, msg, a, b) =
```
Nitpicking: I think that `(loc, msg, seen, latest)` would make the code clearer.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+    Misc.try_finally (fun () ->
+        match Sys.command cmd with
+        | 0 -> Ok (f tmp)
+        | 128 -> Error `Not_found
```
Any specific reason for not using a `exn` payload here, using `Not_found` and `Exit` directly?

edwintorok replied:
exhaustiveness check in the caller.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+    let is_old v =
+      version_of_string v < (4,0,0) || same_version v first
+    in
+    if k = "type Pervasives.format6" then None (* was built-in before *)
```
Again, having this hardcoded constant be defined in a list at the toplevel (and explained) could help people adapt the code more easily.

gasche commented on tools/lintapidiff.sh:
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+set -e
+TAGS=$(git tag|grep '^[0-9]*.[0-9]*.[0-9]*$'|grep -v '^[12].')
+ocamlfind ocamlopt -package compiler-libs.common,str tools/lintapidiff.ml \
```
I see that you documented the assumption that this tool relies on an existing installation with `findlib`, but is there any particular reason for this, or just that you found it easier to iterate on it as a separate tool? I think the reasoning can go both ways: if it's in ` tools` it would be consistent to link with the compiler distribution directly, but having it external like this makes it easier for other people to just steal the scripts and use them for their projects. I like that second advantage and I suspect it could also benefit some other tools, so I would leave things the way you created them.

edwintorok replied:
Initially this had `#use` and `#require` (easier to edit/run from within Emacs). I could move this to the Makefile (and not built it by default, just have a target you can run that builds and runs the linter).

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+
+module Diff = struct
+  type seen_info = {
+    not_seen: version option;
```
Where is `not_seen` set? I understand from the report that it identifies a version in which the item is missing (I would thus find `option list` more natural), but it seems to be `None` everywhere in the code.

edwintorok replied:
In `parse_file_at_rev` `merge`: https://github.com/ocaml/ocaml/pull/916/commits/20e6a66b07a5c27f70b1c2b8846b7e5c6f596cdc#diff-8de361ea648ec4912974450f0d705a9fR193. I think `last_not_seen`, would be a better name.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+    | None, None -> assert false
+    | _, Some {has_doc_parent=false;has_doc=false;_} -> None (* undocumented *)
+    | Some {deprecated=true;first_seen;_}, None -> None (* deleted deprecated *)
+    | Some _, None ->
```
In the case above `{deprecated=true;_}` should suffice. In the present case I think `{deprecated=false;_}` would be clearer -- it's best to avoid relying on case ordering when it is easy.

edwintorok replied:
`first_seen` in `{deprecated=true;first_seen;_}` was required for disambiguation, although I can see how thats a bit ugly, I'll add a type annotation instead.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+    | Some {deprecated=true;first_seen;_}, None -> None (* deleted deprecated *)
+    | Some _, None ->
+        Some (default, "deleted non-deprecated", seen, latest)
+    | _, Some {deprecated=true;since=None;_} -> None (* marked as deprecated *)
```
Isn't it also a problem if deprecated items don't have a `@since` marker?

edwintorok replied:
I think that if something is marked as deprecated users would be better of avoiding to use it, than figuring out the required compiler version constraints where they can use it. e.g. `Array.make_float` was rather short-lived: introduced in 4.02.0, then marked as deprecated, removing the since annotation.

gasche commented on tools/lintapidiff.ml:
```diff
@@ -0,0 +1,268 @@
+    | Some {first_seen;_}, Some {since=None;_} when is_old first_seen ->
+        None (* OK, very old *)
+    | Some {first_seen;_}, Some {loc; since=None;_} ->
+        Some (loc, "missing @since", seen, latest)
```
I think the `when` + case ordering between two clauses is less readable than having a single clause with a conditional test.

gasche reviewed (approved):
I did a review to make sure that I would be comfortable adding the code to the compiler distribution. It is nitpicky, mostly as a way to force myself to read the code. I agree that the design is reasonably solid and I like this implementation approach.

(One disadvantage over the more greppy previous approach is that it won't work very well for code that is not immediately valid OCaml code, eg. if it uses a cpp-like preprocessor, but I think people can adjust their build to run the tool after preprocessing in this case.)

edwintorok commented:
Thanks a lot for the review. I appreciate nitpicky comments, and pushed some commits to address most of them (I haven't replied to each comment individually, but see the commit message in the fixup commits).

shindere commented:
Is there any policy regarding when GPL headers should be added to files?
tools/lintapidiff.ml does not have any at the moment.

gasche commented:
@damiendoligez may know about the policy. @damiendoligez, I would also be interested in your general opinion on the patch. I personally think that it is good and that we should consider merging it.

@edwintorok thanks for the fixups. I agree with your justifications in the no-change cases.

damiendoligez commented:
I like the idea of this patch (both the tool and the fixes) although I don't have time to review (and I trust @gasche for that).

The copyright header is mandatory on all source files. @edwintorok have you signed the CLA already for some other contribution?

edwintorok commented:
@damiendoligez I haven't signed a CLA yet. Can I contact you via e-mail to discuss the details of signing that please?

gasche commented:
@edwintorok it is of course fine to contact Damien if you have further questions, but there is an explanation on the CLA in the [CONTRIBUTING.md](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#contributor-license-agreement) file, both of what it is (and why) and what is the process to complete it (essentially the process is to print a document, sign it, scan it back, and send it to Xavier Leroy by email).

edwintorok commented:
@gasche I have rebased the branch, fixed some new warnings from `lintapidiff`, and added the copyright header on `lintapidiff.ml`. Please let me know if these changes are ok, I'll then squash the commits marked with `squash!`

xavierleroy commented:
CLA received in good order.  I don't know what the next step is, but if anyone does, please proceed.

gasche commented:
I think this should be merged in `trunk` and `4.05`, and I can take care of it. @edwintorok, could you squash and also add a Changes entry (this was a sizeable amount of work so it deserves one)?

edwintorok commented:
Thanks, squashed it down to 2 commits and added the Changes entry.

gasche commented:
(Hm, I was only thinking of squashing the `squash!` commit and you squashed a bit more. That's fine.)

The commit message in 640bc8bcac9b306c33e1aa8791f8b4b45575f0b3 and the comment in `lintapidiff.ml` still mention a `lintapidiff.sh` script, but as far as I can tell this script does not exist in the current version of the patch.

edwintorok commented:
Thanks for spotting that, pushed updated branch.

gasche commented:
Merged, and pushed to 4.05 as well. Thanks!

gasche commented:
Note: we just used `lintapidiff` in preparation for the 4.06, and it was useful (we caught a missing `@since` in stdlib/uchar.mli). Thanks again!
